ISSUE-910 Test Mode: Getting an exception on deleting a test case tha…

HeartSaVioR opened the pull request:
…t has been ran successfully once

* also name all foreign keys via dropping all and recreating
* drop FK of topology test case on topology test history

This closes #910 and maybe also #911 too.

As I stated from #910, we added foreign key without naming symbol, so that's addressed for v004. I resolved origin issue from v005.

Tested both MySQL and PostgreSQL, but for PostgreSQL I had to fix some table schemas since some foreign keys are not defined as same as others.

@shahsank3t Please apply and see if it resolves #910. Please let me know when you have hard time to apply database migration.

@arunmahadevan @raju-saravanan Please take a look and comment. Thanks in advance.

arunmahadevan commented on bootstrap/sql/mysql/v005__drop_test_case_fk_from_topology_test_history.sql:
```diff
@@ -0,0 +1,16 @@
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+ALTER TABLE `topology_test_run_histories` DROP FOREIGN KEY `fk_topology_test_run_histories_topology_test_run_case`, DROP KEY `fk_topology_test_run_histories_topology_test_run_case`;
```
Do we need this if we dont add the FK constraint it in the above script?

HeartSaVioR replied:
Yes I broke down file because of explicitly showing intention (like breaking down commits) but we could address it to v004 and leave comment explaining it. I'll address it.

arunmahadevan commented on bootstrap/sql/postgresql/v001__create_tables.sql:
```diff
@@ -51,8 +52,9 @@ CREATE TABLE IF NOT EXISTS datasource (
   "timestamp"  BIGINT,
-  "dashboardId"  INTEGER REFERENCES  dashboard,
-  CONSTRAINT datasource_uk_name UNIQUE ("name")
+  "dashboardId" BIGINT NOT NULL,
```
Any changes to within "create table if not exists" will not be applied during migration in the next release version. Do we want to put these changes in a separate script?

HeartSaVioR replied:
Ideally yes, but could we decide the base version first? I think we already made a change in v001 so if then we should also fix that.

HeartSaVioR replied:
Realized the change is not needed. Rolled back.

arunmahadevan commented on bootstrap/sql/postgresql/v005__drop_test_case_fk_from_topology_test_history.sql:
```diff
@@ -0,0 +1,15 @@
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+ALTER TABLE "topology_test_run_histories" DROP CONSTRAINT IF EXISTS "fk_topology_test_run_histories_topology_test_run_case";
```
do we need if we dont add it in v004?

HeartSaVioR replied:
Same here. I'll apply this to v003.

arunmahadevan commented:
@HeartSaVioR thanks for the patch. Looks good.

@raju-saravanan can you take a look? While reviewing also see if there are any other cases (PK or any other constraints) that we need to cover and after that you can close https://github.com/hortonworks/streamline/issues/911

raju-saravanan commented on bootstrap/sql/postgresql/v001__create_tables.sql:
```diff
@@ -39,8 +39,9 @@ CREATE TABLE IF NOT EXISTS widget (
   "type" VARCHAR(255) NOT NULL,
   "data" TEXT NOT NULL,
   "timestamp"  BIGINT,
+  "dashboardId" BIGINT NOT NULL,
```
If the user already has "widget" entity in this database, running this migration won't reflect changes to the column "dashboardId" as the table is created with "create if not exist" syntax. Can we add another sql migration script to modify dashboardId ?

HeartSaVioR replied:
Yes you're right. Maybe need to have one another sql for postgre or fix them all from v004.

raju-saravanan commented on bootstrap/sql/postgresql/v001__create_tables.sql:
```diff
@@ -39,8 +39,9 @@ CREATE TABLE IF NOT EXISTS widget (
+  "dashboardId" BIGINT NOT NULL,
   CONSTRAINT widget_uk_name UNIQUE ("name"),
-  "dashboardId"  INTEGER REFERENCES dashboard
+  FOREIGN KEY ("dashboardId") REFERENCES dashboard("id")
```
I think foreign key constraint has to be captured in a separate script as mentioned above

raju-saravanan commented on bootstrap/sql/postgresql/v004__set_name_to_all_foreign_keys.sql:
```diff
@@ -0,0 +1,90 @@
+ALTER TABLE topology_sink ADD CONSTRAINT "fk_topology_sink_topology_version" FOREIGN KEY ("versionId") REFERENCES topology_version("id");
+ALTER TABLE topology_processor ADD CONSTRAINT "fk_topology_processor_topology_version" FOREIGN KEY ("versionId") REFERENCES topology_version("id");
+ALTER TABLE topology_processor_stream_mapping ADD CONSTRAINT "fk_topology_processor_stream_mapping_topology_processor" FOREIGN KEY ("processorId", "versionId") REFERENCES topology_processor("id", "versionId");
+ALTER TABLE topology_edge ADD CONSTRAINT "fk_topology_edge_topology_version" FOREIGN KEY ("versionId") REFERENCES topology_version("id");
```
Seems like fk constraint  "FOREIGN KEY (streamId, versionId) REFERENCES topology_stream(id, versionId)" for topology_processor_stream_mapping is missing here.

HeartSaVioR replied:
Nice finding. Will add.

HeartSaVioR commented:
@arunmahadevan @raju-saravanan 
Addressed review comments. 
Btw, as I commented earlier, I think we already made a change in v001 , so we have to decide the base version first, and need to sort out.

arunmahadevan commented:
@raju-saravanan , can you move any changes made in v001 since HDF-3.0.0.0 and add it to a separate scripts ? 

i.e. use the HDF-3.0.0.0 branch as the baseline for ./bootstrap/sql/mysql/v001__create_tables.sql

Any new changes should go to new scripts. Without that the migrations would fail.

We also made minor release after 3.0.0, so I am not sure how would it work if someone migrates from that minor release to HDF-3.1.

arunmahadevan commented on bootstrap/sql/postgresql/v003__add_sleep_iteration_to_topology_test_run_case_source.sql:
```diff
@@ -13,4 +13,4 @@
 -- ISSUE-902: Test mode: Add a sleep option along with repeat count
-ALTER TABLE `topology_test_run_case_source` ADD COLUMN sleepMsPerIteration BIGINT NOT NULL DEFAULT 0;
\ No newline at end of file
+ALTER TABLE "topology_test_run_case_source" ADD COLUMN "sleepMsPerIteration" BIGINT NOT NULL DEFAULT 0;
```
is this not required for mysql ?

HeartSaVioR replied:
That was available for both mysql and postgresql, but I realized that postgresql needs `"`, so I fixed that. MySQL is OK.

arunmahadevan commented:
+1
